src/ds/logger.h

```diff
@@ -29,6 +31,136 @@ namespace logger
 
   static constexpr size_t ns_per_s = 1'000'000'000;
 
+  class CustomLogger
+  {
+  protected:
+    std::string log_path;
+
+  public:
+    CustomLogger() = default;
+    CustomLogger(std::string log_path_) : log_path(log_path_) {}
+    virtual ~CustomLogger() = default;
+
+    std::string get_timestamp(const std::tm& tm, const ::timespec& ts)
+    {
+      // Sample: "2019-07-19 18:53:25.690267"
+      return fmt::format("{:%Y-%m-%d %H:%M:%S}.{:0<6}", tm, ts.tv_nsec / 1000);
+    }
+
+    virtual std::string format(
+      const std::string& file_name,
+      size_t line_number,
+      const std::string& log_level,
+      const std::string& msg,
+      const std::tm& host_tm,
+      const ::timespec& host_ts,
+      const std::optional<::timespec>& enclave_ts = std::nullopt) = 0;
+
+    virtual void write(const std::string& log_line) = 0;
+
+    void dump(const std::string& msg, const std::string& file)
+    {
+      using namespace std;
+      ofstream f(file, ios_base::app);
```

Comment:
Minor: Do we need to open this `ofstream` each time, or can we make this the member replacing `log_path`?

Reply:
good idea - done